src/v0/helpers.ts

```diff
@@ -0,0 +1,229 @@
+import { BigInt, Address } from "@graphprotocol/graph-ts";
+import {
+  Steward,
+  LogBuy,
+  LogPriceChange,
+  LogForeclosure,
+  LogCollection,
+  LogRemainingDepositUpdate,
+  AddToken
+} from "../../generated/Steward/Steward";
+import {
+  Wildcard,
+  Patron,
+  PreviousPatron,
+  Price,
+  TokenUri,
+  BuyEvent,
+  EventCounter,
+  ChangePriceEvent,
+  Global
+} from "../../generated/schema";
+import { Token } from "../../generated/Token/Token";
+import { log } from "@graphprotocol/graph-ts";
+import {
+  ZERO_ADDRESS,
+  NUM_SECONDS_IN_YEAR,
+  AMOUNT_RAISED_BY_VITALIK_VINTAGE_CONTRACT,
+  NUM_SECONDS_IN_YEAR_BIG_INT,
+  BILLION,
+  VITALIK_PATRONAGE_NUMERATOR,
+  VITALIK_PATRONAGE_DENOMINATOR,
+  GLOBAL_PATRONAGE_DENOMINATOR
+} from "../CONSTANTS";
+import { getForeclosureTimeSafe, minBigInt } from "../util";
+import {
+  getTotalCollectedAccurate,
+  getTotalOwedAccurate,
+  getTotalTokenCostScaledNumerator
+} from "../util/hacky";
+
+// A token would need to be set to the same price
+export function getTokenIdFromTxTokenPrice(
+  steward: Steward,
+  tokenPrice: BigInt,
+  owner: Address,
+  timestamp: BigInt
+): i32 {
+  // if ("0x6b90b98733928136cb44675929ca5359d1a76417" == owner.toHexString()) {
+  //   log.warning(
+  //     "Matching Marvin's v0 token! timestamp: {} == {} ? \n ? tokenPrice: {} == {} ? \n ? owner: {} == {}",
+  //     [
+  //       timestamp.toString(),
+  //       steward.timeLastCollected(BigInt.fromI32(0)).toString(),
+  //       tokenPrice.toString(),
+  //       steward.price(BigInt.fromI32(0)).toString(),
+  //       owner.toString(),
+  //       steward.currentPatron(BigInt.fromI32(0)).toString()
+  //     ]
+  //   );
+  // }
+  if (
+    timestamp.equals(steward.timeLastCollected(BigInt.fromI32(0))) &&
```

Comment:
There is a mismatch here I think when a foreclosure has happened. We are passing in a timestamp, and checking timestamp.equals(steward.timeLastCollected(BigInt.fromI32(0)))
but when a foreclsoure happens, this code in the smart contract executes:

if (patronageOwedByTokenPatron >= deposit[currentOwner]) {
                uint256 newTimeLastCollected = timeLastCollectedPatron[currentOwner]
                    .add(
                    (
                        (now.sub(timeLastCollectedPatron[currentOwner]))
                            .mul(deposit[currentOwner])
                            .div(patronageOwedByTokenPatron)
                    )
                );

                timeLastCollected[tokenId] = newTimeLastCollected;

So in this case the timestamp of the wont equal timeLastCollected, as timeLastCollected is adjusted for the foreclosure?

Maybe I'm wrong on this just want to be sure

Reply:
Yeah, you're right! Good catch. But I don't think any tokens foreclosed on that contract. So we are safe :+1: 

In the future to be very very safe we can have a 

```
switch(event.transaction.hash.toHexString()) {
case("0x123):
 return 3
case("0xabc):
 return 6
...
...
}
```